Add Rust package rage

fhoehle opened the pull request:
<!--
Thank you very much for putting in this recipe PR!

This repository is very active, so if you need help with
a PR or once it's ready for review, please let the right people know.
There are language-specific teams for reviewing recipes.

Currently available teams are:
- python `@conda-forge/help-python`
- python/c hybrid `@conda-forge/help-python-c`
- r `@conda-forge/help-r`
- java `@conda-forge/help-java`
- nodejs `@conda-forge/help-nodejs`
- c/c++ `@conda-forge/help-c-cpp`
- perl `@conda-forge/help-perl`
- Julia `@conda-forge/help-julia`
- ruby `@conda-forge/help-ruby`

If your PR doesn't fall into those categories please contact
the full review team `@conda-forge/staged-recipes`.

Due to GitHub limitations first time contributors to conda-forge are unable
to ping these teams. You can [ping the team](https://conda-forge.org/docs/maintainer/infrastructure.html#conda-forge-admin-please-ping-team) using a special command in
a comment on the PR to get the attention of the `staged-recipes` team. You can
also consider asking on our [Gitter channel](https://gitter.im/conda-forge/conda-forge.github.io)
or on our [Keybase chat](https://keybase.io/team/condaforge.chat)
if your recipe isn't reviewed promptly.
-->

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml"
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example)
- [x] Source is from official source
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged)
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details)
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/rage```) and found it was in an excellent condition.

fhoehle commented:
@conda-forge/staged-recipes please be aware of the license issues of Rust packages: https://github.com/conda-forge/conda-forge.github.io/issues/1052 I tried to follow the given guidlines and tried to track down every license of all given depencies. The package `bech32` is [excluded](https://github.com/conda-forge/staged-recipes/blob/35eb2c6f9c20cadf3901db929f84add272790b54/recipes/rage/check_licenses.py#L18) from license checking since I was not able to find any. The [`cargo.toml`](https://crates.io/crates/bech32) is listing it as MIT but at its [git](https://github.com/rust-bitcoin/rust-bech32). Should be an empty file included?

isuruf commented:
See https://github.com/rust-bitcoin/rust-bech32/blob/master/src/lib.rs#L1-L19

fhoehle commented:
> See https://github.com/rust-bitcoin/rust-bech32/blob/master/src/lib.rs#L1-L19

Thanks, is added.

fhoehle commented:
@conda-forge/staged-recipes: This is ready to review or comment.

In order to facilitate the review, below a procedure to obtain the license file(s) of all dependencies.

- Which license is prefered if more than one is available?
  - If available the MIT license is added here.
  - If no MIT is available APACHE is selected.
- Here scripts to assure all neccessary licenses were added:

    1. Getting all dependencies from [`rage-0.4.0.tar.gz`](https://github.com/str4d/rage/archive/v0.4.0.tar.gz) defined in `rage/Cargo.lock` using `cd rage && cargo license -j > dependencies.json`
    2. Running following Python script which downloads all crates except the modules provided by the `tgz` archive (`age`, `age-core`, `rage`):
        <details>
 
        ```python
        import json, requests, tarfile, gzip, io, itertools, re, io, os

        licenses = json.load(open("dependencies.json"))
        licenses_and_crate={};
        skipping = ["rage","age","age-core"] 
        for i, l in enumerate(licenses):
            name, version = l["name"], l["version"]
            if name in skipping:
                continue
            print(name, version, f"{i+1}/{len(licenses)}")
            res = requests.get(f"https://crates.io/api/v1/crates/{name}/{version}/download")
            tf = tarfile.TarFile(fileobj=gzip.GzipFile(fileobj=io.BytesIO(res.content)),mode='r')
            l["tarfile"] = tf
            licenses_and_crate[f"{l['name']}-{l['version']}"] = l

        assert len(licenses_and_crate) + len(skipping) == len(licenses)

        custom={'ryu-1.0.3': 'LICENSE-APACHE', "ansi_term-0.12.1": "LICENCE", "minreq-2.1.0": "COPYING.md"}

        explicit_mit_licenses = []; license_file = []; missing=[];
        for k,l in licenses_and_crate.items():
            tf = l["tarfile"]
            files = [ti.name.split('/')[-1] for ti in tf.getmembers() if len(ti.name.split('/')) == 2]
            if 'LICENSE-MIT' in files:
                explicit_mit_licenses.append((f"{l['name']}",f"{l['version']}",l["license"],'LICENSE-MIT'))
            else:
                matches = [ f for f in files if re.match("LICENSE*",f)]
                if len(matches) ==1:
                    license_file.append((f"{l['name']}",f"{l['version']}",l["license"],matches[0]))
                else:
                    if k in custom:
                        license_file.append((f"{l['name']}",f"{l['version']}",l["license"],custom[k]))
                    else:
                        missing.append((f"{l['name']}",f"{l['version']}",l["license"],l['repository']))

        print("\nmissing libray licenses:")
        print("\n".join([str(l) for l in missing]))
        print(f"{len(explicit_mit_licenses)} having 'LICENSE-MIT', {len(license_file)} having 'LICENSE*' or custom filename, {len(missing)} licenses are missing!!!; in total {len(licenses_and_crate)}")

        tmp_dir = "./tmp_licenses"
        os.makedirs(tmp_dir, exist_ok=True)
        for name, version, _, lic_name in itertools.chain(explicit_mit_licenses, license_file):
            key = f"{name}-{version}"
            tf = licenses_and_crate[key]["tarfile"]
            with open(os.path.join(tmp_dir, f"{key}-license"), "wb") as f:
                f.write(tf.extractfile(f"{key}/{lic_name}").read().replace(b'\r\n', b'\n'))
        ```

        </details>

        yields: 

        ```
        adler32 1.0.4 1/179
        ...
        zeroize_derive 1.0.0 179/179

        missing libray licenses:
        ('bech32', '0.7.2', 'MIT', 'https://github.com/rust-bitcoin/rust-bech32')
        ('blake2b_simd', '0.5.10', 'MIT', 'https://github.com/oconnor663/blake2_simd')
        ('clap', '3.0.0-beta.1', 'Apache-2.0 OR MIT', 'https://github.com/clap-rs/clap')
        ('clap_derive', '3.0.0-beta.1', 'Apache-2.0 OR MIT', 'https://github.com/clap-rs/clap/tree/master/clap_derive')
        ('clap_generate', '3.0.0-beta.1', 'Apache-2.0 OR MIT', 'https://github.com/clap-rs/clap/tree/master/clap_generate')
        ('clicolors-control', '1.0.1', 'MIT', 'https://github.com/mitsuhiko/clicolors-control')
        ('criterion-cycles-per-byte', '0.1.1', 'Apache-2.0 OR MIT', 'https://gitlab.com/sio4/code/criterion-cycles-per-byte')
        ('gumdrop_derive', '0.7.0', 'Apache-2.0 OR MIT', 'https://github.com/murarth/gumdrop')
        ('quickcheck_macros', '0.9.1', 'MIT OR Unlicense', 'https://github.com/BurntSushi/quickcheck')
        ('winapi-i686-pc-windows-gnu', '0.4.0', 'Apache-2.0 OR MIT', 'https://github.com/retep998/winapi-rs')
        ('winapi-x86_64-pc-windows-gnu', '0.4.0', 'Apache-2.0 OR MIT', 'https://github.com/retep998/winapi-rs')
        132 having 'LICENSE-MIT', 33 having 'LICENSE*' or custom filename, 11 licenses are missing!!!; in total 176
        ```
    3. Prooving same files are given here: 
        ```
        diff -r tmp_licenses/ library_licenses/
        ```
        yields:
        ```
        Only in library_licenses/: bech32-0.7.2-license
        Only in library_licenses/: blake2b_simd-0.5.10-license
        Only in library_licenses/: clap-3.0.0-beta.1-license
        Only in library_licenses/: clap_derive-3.0.0-beta.1-license
        Only in library_licenses/: clap_generate-3.0.0-beta.1-license
        Only in library_licenses/: clicolors-control-1.0.1-license
        Only in library_licenses/: criterion-cycles-per-byte-0.1.1-license
        Only in library_licenses/: gumdrop_derive-0.7.0-license
        Only in library_licenses/: quickcheck_macros-0.9.1-license
        Only in library_licenses/: winapi-i686-pc-windows-gnu-0.4.0-license
        Only in library_licenses/: winapi-x86_64-pc-windows-gnu-0.4.0-license
        ```
    4. The missing 11 license files were downloaded manually.
- Summary of all dependencies and their licenses:
   ```
    adler32 1.0.4: Zlib
    aead 0.2.0: Apache-2.0 OR MIT
    aes 0.3.2: Apache-2.0 OR MIT
    aes-ctr 0.3.0: Apache-2.0 OR MIT
    aes-soft 0.3.3: Apache-2.0 OR MIT
    aesni 0.6.0: Apache-2.0 OR MIT
    age 0.4.0: Apache-2.0 OR MIT
    age-core 0.4.0: Apache-2.0 OR MIT
    aho-corasick 0.7.10: MIT OR Unlicense
    ansi_term 0.12.1: MIT
    arrayref 0.3.6: BSD-2-Clause
    arrayvec 0.4.12: Apache-2.0 OR MIT
    arrayvec 0.5.1: Apache-2.0 OR MIT
    atty 0.2.14: MIT
    autocfg 1.0.0: Apache-2.0 OR MIT
    base64 0.10.1: Apache-2.0 OR MIT
    base64 0.11.0: Apache-2.0 OR MIT
    bcrypt-pbkdf 0.1.0: Apache-2.0 OR MIT
    bech32 0.7.2: MIT
    bitflags 1.2.1: Apache-2.0 OR MIT
    blake2b_simd 0.5.10: MIT
    block-buffer 0.7.3: Apache-2.0 OR MIT
    block-cipher-trait 0.6.2: Apache-2.0 OR MIT
    block-modes 0.3.3: Apache-2.0 OR MIT
    block-padding 0.1.5: Apache-2.0 OR MIT
    blowfish 0.4.0: Apache-2.0 OR MIT
    bstr 0.2.12: Apache-2.0 OR MIT
    bumpalo 3.2.1: Apache-2.0 OR MIT
    byte-tools 0.3.1: Apache-2.0 OR MIT
    byteorder 1.3.4: MIT OR Unlicense
    c2-chacha 0.2.3: Apache-2.0 OR MIT
    cast 0.2.3: Apache-2.0 OR MIT
    cc 1.0.50: Apache-2.0 OR MIT
    cfg-if 0.1.10: Apache-2.0 OR MIT
    chacha20poly1305 0.4.1: Apache-2.0 OR MIT
    chrono 0.4.11: Apache-2.0 OR MIT
    clap 2.33.0: MIT
    clap 3.0.0-beta.1: Apache-2.0 OR MIT
    clap_derive 3.0.0-beta.1: Apache-2.0 OR MIT
    clap_generate 3.0.0-beta.1: Apache-2.0 OR MIT
    clicolors-control 1.0.1: MIT
    console 0.10.0: MIT
    constant_time_eq 0.1.5: CC0-1.0
    cookie-factory 0.3.1: MIT
    crc32fast 1.2.0: Apache-2.0 OR MIT
    criterion 0.3.1: Apache-2.0 OR MIT
    criterion-cycles-per-byte 0.1.1: Apache-2.0 OR MIT
    criterion-plot 0.4.1: Apache-2.0 OR MIT
    crossbeam-deque 0.7.3: Apache-2.0 OR MIT
    crossbeam-epoch 0.8.2: Apache-2.0 OR MIT
    crossbeam-queue 0.2.1: Apache-2.0 AND BSD-2-Clause OR MIT
    crossbeam-utils 0.7.2: Apache-2.0 OR MIT
    crypto-mac 0.7.0: Apache-2.0 OR MIT
    csv 1.1.3: MIT OR Unlicense
    csv-core 0.1.10: MIT OR Unlicense
    ctr 0.3.2: Apache-2.0 OR MIT
    curve25519-dalek 2.0.0: BSD-3-Clause
    digest 0.8.1: Apache-2.0 OR MIT
    dirs 2.0.2: Apache-2.0 OR MIT
    dirs-sys 0.3.4: Apache-2.0 OR MIT
    either 1.5.3: Apache-2.0 OR MIT
    encode_unicode 0.3.6: Apache-2.0 OR MIT
    env_logger 0.7.1: Apache-2.0 OR MIT
    fake-simd 0.1.2: Apache-2.0 OR MIT
    flate2 1.0.14: Apache-2.0 OR MIT
    generic-array 0.12.3: MIT
    getrandom 0.1.14: Apache-2.0 OR MIT
    gumdrop 0.7.0: Apache-2.0 OR MIT
    gumdrop_derive 0.7.0: Apache-2.0 OR MIT
    heck 0.3.1: Apache-2.0 OR MIT
    hermit-abi 0.1.8: Apache-2.0 OR MIT
    hkdf 0.8.0: Apache-2.0 OR MIT
    hmac 0.7.1: Apache-2.0 OR MIT
    humantime 1.3.0: Apache-2.0 OR MIT
    indexmap 1.3.2: Apache-2.0 OR MIT
    itertools 0.8.2: Apache-2.0 OR MIT
    itoa 0.4.5: Apache-2.0 OR MIT
    js-sys 0.3.36: Apache-2.0 OR MIT
    lazy_static 1.4.0: Apache-2.0 OR MIT
    lexical-core 0.6.2: Apache-2.0 OR MIT
    libc 0.2.68: Apache-2.0 OR MIT
    log 0.4.8: Apache-2.0 OR MIT
    man 0.3.0: Apache-2.0 OR MIT
    maybe-uninit 2.0.0: Apache-2.0 OR MIT
    memchr 2.3.3: MIT OR Unlicense
    memoffset 0.5.4: MIT
    miniz_oxide 0.3.6: MIT
    minreq 2.1.0: ISC
    nodrop 0.1.14: Apache-2.0 OR MIT
    nom 5.1.1: MIT
    num-integer 0.1.42: Apache-2.0 OR MIT
    num-traits 0.2.11: Apache-2.0 OR MIT
    num_cpus 1.12.0: Apache-2.0 OR MIT
    oorandom 11.1.0: MIT
    opaque-debug 0.2.3: Apache-2.0 OR MIT
    pbkdf2 0.3.0: Apache-2.0 OR MIT
    pinentry 0.1.0: Apache-2.0 OR MIT
    plotters 0.2.12: MIT
    poly1305 0.5.2: Apache-2.0 OR MIT
    ppv-lite86 0.2.6: Apache-2.0 OR MIT
    proc-macro-error 0.4.12: Apache-2.0 OR MIT
    proc-macro-error-attr 0.4.12: Apache-2.0 OR MIT
    proc-macro2 1.0.9: Apache-2.0 OR MIT
    quick-error 1.2.3: Apache-2.0 OR MIT
    quickcheck 0.9.2: MIT OR Unlicense
    quickcheck_macros 0.9.1: MIT OR Unlicense
    quote 1.0.3: Apache-2.0 OR MIT
    radix64 0.6.2: Apache-2.0 OR MIT
    rage 0.4.0: Apache-2.0 OR MIT
    rand 0.7.3: Apache-2.0 OR MIT
    rand_chacha 0.2.2: Apache-2.0 OR MIT
    rand_core 0.5.1: Apache-2.0 OR MIT
    rand_hc 0.2.0: Apache-2.0 OR MIT
    rayon 1.3.0: Apache-2.0 OR MIT
    rayon-core 1.7.0: Apache-2.0 OR MIT
    redox_syscall 0.1.56: MIT
    redox_users 0.3.4: MIT
    regex 1.3.5: Apache-2.0 OR MIT
    regex-automata 0.1.9: MIT OR Unlicense
    regex-syntax 0.6.17: Apache-2.0 OR MIT
    ring 0.16.11: None
    roff 0.1.0: Apache-2.0 OR MIT
    rpassword 4.0.5: Apache-2.0
    rust-argon2 0.7.0: Apache-2.0 OR MIT
    rustc_version 0.2.3: Apache-2.0 OR MIT
    rustls 0.16.0: Apache-2.0 OR ISC OR MIT
    ryu 1.0.3: Apache-2.0 OR BSL-1.0
    same-file 1.0.6: MIT OR Unlicense
    scopeguard 1.1.0: Apache-2.0 OR MIT
    scrypt 0.2.0: Apache-2.0 OR MIT
    sct 0.6.0: Apache-2.0 OR ISC OR MIT
    secrecy 0.6.0: Apache-2.0 OR MIT
    semver 0.9.0: Apache-2.0 OR MIT
    semver-parser 0.7.0: Apache-2.0 OR MIT
    serde 1.0.105: Apache-2.0 OR MIT
    serde_derive 1.0.105: Apache-2.0 OR MIT
    serde_json 1.0.48: Apache-2.0 OR MIT
    sha2 0.8.1: Apache-2.0 OR MIT
    spin 0.5.2: MIT
    static_assertions 0.3.4: Apache-2.0 OR MIT
    stream-cipher 0.3.2: Apache-2.0 OR MIT
    strsim 0.9.3: MIT
    subtle 1.0.0: BSD-3-Clause
    subtle 2.2.2: BSD-3-Clause
    syn 1.0.17: Apache-2.0 OR MIT
    syn-mid 0.5.0: Apache-2.0 OR MIT
    synstructure 0.12.3: MIT
    termcolor 1.1.0: MIT OR Unlicense
    termios 0.3.1: MIT
    textwrap 0.11.0: MIT
    thread_local 1.0.1: Apache-2.0 OR MIT
    time 0.1.42: Apache-2.0 OR MIT
    tinytemplate 1.0.3: Apache-2.0 OR MIT
    typenum 1.11.2: Apache-2.0 OR MIT
    unicode-segmentation 1.6.0: Apache-2.0 OR MIT
    unicode-width 0.1.7: Apache-2.0 OR MIT
    unicode-xid 0.2.0: Apache-2.0 OR MIT
    universal-hash 0.3.0: Apache-2.0 OR MIT
    untrusted 0.7.0: ISC
    vec_map 0.8.1: Apache-2.0 OR MIT
    version_check 0.9.1: Apache-2.0 OR MIT
    walkdir 2.3.1: MIT OR Unlicense
    wasi 0.9.0+wasi-snapshot-preview1: Apache-2.0 OR Apache-2.0 WITH LLVM-exception OR MIT
    wasm-bindgen 0.2.59: Apache-2.0 OR MIT
    wasm-bindgen-backend 0.2.59: Apache-2.0 OR MIT
    wasm-bindgen-macro 0.2.59: Apache-2.0 OR MIT
    wasm-bindgen-macro-support 0.2.59: Apache-2.0 OR MIT
    wasm-bindgen-shared 0.2.59: Apache-2.0 OR MIT
    web-sys 0.3.36: Apache-2.0 OR MIT
    webpki 0.21.2: None
    webpki-roots 0.18.0: MPL-2.0
    which 3.1.1: MIT
    winapi 0.3.8: Apache-2.0 OR MIT
    winapi-i686-pc-windows-gnu 0.4.0: Apache-2.0 OR MIT
    winapi-util 0.1.3: MIT OR Unlicense
    winapi-x86_64-pc-windows-gnu 0.4.0: Apache-2.0 OR MIT
    x25519-dalek 0.6.0: BSD-3-Clause
    zeroize 1.1.0: Apache-2.0 OR MIT
    zeroize_derive 1.0.0: Apache-2.0 OR MIT
    ```

isuruf commented on recipes/rage/meta.yaml:
```diff
@@ -0,0 +1,32 @@
+{% set version = "v0.4.0" %}
```
```suggestion
{% set version = "0.4.0" %}
```

isuruf commented on recipes/rage/meta.yaml:
```diff
@@ -0,0 +1,32 @@
+  name: rage
+  version: {{ version.lstrip('v') }}
+source:
+  url: https://github.com/str4d/rage/archive/{{ version }}.tar.gz
```
```suggestion
  url: https://github.com/str4d/rage/archive/v{{ version }}.tar.gz
```
